sycl/source/detail/persistent_cache.cpp

```diff
@@ -0,0 +1,259 @@
+//==---------- persistent_cache.cpp - On-disk cache for program -*- C++-*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <detail/config.hpp>
+#include <detail/persistent_cache.hpp>
+#include <detail/plugin.hpp>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+/* Stores build program in persisten cache
+ */
+void PersistentCache::putPIProgramToDisc(const detail::plugin &Plugin,
+                                         const device &Device,
+                                         const RTDeviceBinaryImage &Img,
+                                         const SerializedObj &SpecConsts,
+                                         const std::string &BuildOptionsString,
+                                         const RT::PiProgram &Program) {
+
+  if (!isPersistentCacheEnabled())
+    return;
+
+  // Only SPIRV images are cached
+  if (Img.getFormat() != PI_DEVICE_BINARY_TYPE_SPIRV &&
+      (Img.getFormat() == PI_DEVICE_BINARY_TYPE_NONE &&
+       pi::getBinaryImageFormat(Img.getRawData().BinaryStart, Img.getSize()) !=
+           PI_DEVICE_BINARY_TYPE_SPIRV))
+    return;
+
+  std::string DirName =
+      getCacheItemDirName(Device, Img, SpecConsts, BuildOptionsString);
+
+  size_t i = 0;
+  std::string FileName;
```

Comment:
I think this is an important part which should be documented (if it is already documented in the design doc, then some summary here would be really helpful) - what files constitute the cache, what are their content, what is name generation algo, how they are found upon subsequent runs to read particular program from.